### PR TITLE
Cherry-pick of "Selective merged prefill #643"

### PR DESCRIPTION
**Pull request by kamil-kaczor**
Cherry-pick of https://github.com/HabanaAI/vllm-fork/pull/643

### CODE REVIEW

File: vllm/worker/hpu_model_runner.py [hunk 1]
```diff
@@ -227,6 +228,33 @@ def find_rope_layer(parent, path):
     return path_to_rope
 
 
+class HPUBucketingContextWithMergedPrefill(HPUBucketingContext):
```
Comment (kdamaszk) [inline vllm/worker/hpu_model_runner.py:231]: I think this should be a part of https://github.com/HabanaAI/vllm-hpu-extension/blob/main/vllm_hpu_extension/bucketing.py

File: vllm/worker/hpu_model_runner.py [hunk 2]
```diff
@@ -705,10 +773,16 @@ def __init__(
         self.profiler_counter_helper = HabanaProfilerCounterHelper()
         self.seen_configs: set = set()
         self._mem_margin: Optional[int] = None
-        self.bucketing_ctx = HPUBucketingContext(self.max_num_seqs,
-                                                 self.max_num_prefill_seqs,
-                                                 self.block_size,
-                                                 self.max_num_batched_tokens)
+        self.enable_merged_prefill = os.environ.get('VLLM_MERGED_PREFILL',
```
Comment (kdamaszk) [inline vllm/worker/hpu_model_runner.py:776]: It would be good to add this flag to the README with small explanation.

File: vllm/worker/hpu_model_runner.py (hunk 1 above)
Comment (madamczykhabana) [reply vllm/worker/hpu_model_runner.py:231]: +1

File: vllm/worker/hpu_model_runner.py [hunk 3]
```diff
@@ -1717,8 +2044,11 @@ def profile_run(self) -> None:
         max_batch_size = min(self.max_num_seqs,
                              self.max_num_batched_tokens // max_seq_len)
 
+        origin_enable_merged_prefill = self.enable_merged_prefill
+        self.enable_merged_prefill = False
         self.warmup_scenario(max_batch_size, max_seq_len, True, kv_caches,
                              False, True)
+        self.enable_merged_prefill = origin_enable_merged_prefill
```
Comment (madamczykhabana) [inline vllm/worker/hpu_model_runner.py:2051]: This is a code smell. Could we do it differently?

File: vllm/worker/hpu_model_runner.py [hunk 4]
```diff
@@ -1189,6 +1263,241 @@ def _prepare_prompt(
...
+                    block_table = seq_group_metadata.block_tables[seq_id]
+                    prefix_block_tables.append(block_table)
+                else:
+                    # The first prefill.
+                    prefix_block_tables.append([])
+            else:
+                prefix_block_tables.append([])
+                # Right now, prefill start is always 0. However, this
+                # assumption can be changed once chunked prefill is introduced.
+                assert context_len == 0
```
Comment (kzawora-intel) [inline vllm/worker/hpu_model_runner.py:1329]: do we support prefix caching in merged prefill?

File: vllm/worker/hpu_model_runner.py [hunk 5]
```diff
@@ -1189,6 +1263,241 @@ def _prepare_prompt(
...
+        slot_mapping = make_tensor_with_pad(slot_mapping,
+                                            max_len=merged_prompt_len,
+                                            pad=_PAD_SLOT_ID,
+                                            dtype=torch.long,
+                                            device='cpu')
+        actual_num_prefills_tensor = torch.tensor(len(seq_lens),
+                                                  dtype=torch.long,
+                                                  device='cpu')
+
+        max_prefill_bs = int(os.environ.get('VLLM_PROMPT_BS_BUCKET_MAX', '8'))
```
Comment (kzawora-intel) [inline vllm/worker/hpu_model_runner.py:1429]: we already have a parameter in scheduler_config for that, max_num_prefill_seqs

File: vllm/worker/hpu_model_runner.py [hunk 6]
```diff
@@ -1189,6 +1263,241 @@ def _prepare_prompt(
...
+            seq_id = seq_ids[0]
+
+            computed_block_nums = seq_group_metadata.computed_block_nums
+            if (self.scheduler_config is not None
+                    and self.scheduler_config.chunked_prefill_enabled
+                    and not (computed_block_nums is None
+                             or computed_block_nums == [])):
+                raise RuntimeError(
+                    "chunked prefill cannot be used with prefix caching "
+                    "now.")
```
Comment (kzawora-intel) [inline vllm/worker/hpu_model_runner.py:1299]: we don't support chunked prefill at all

Comment (michalkuligowski) [reply vllm/worker/hpu_model_runner.py:1299]: That is correct, there is this PR however https://github.com/HabanaAI/vllm-hpu-extension/pull/94/

File: vllm/attention/backends/hpu_attn.py [hunk 7]
```diff
@@ -24,6 +24,38 @@
...
+    query = query.transpose(1, 2)
+    key = key.transpose(1, 2)
+    value = value.transpose(1, 2)
+    softmax_mode = 'fast'
+    recompute_mode = True
+    attn_weights = fsdpa_op(query, key, value, attn_bias, 0.0, False, scale,
+                            softmax_mode, recompute_mode, None, 'right')
+    attn_weights = attn_weights.transpose(1, 2)
+    return attn_weights
+
```
[the remark this answers is not in the document]
Comment (xuechendi) [reply vllm/attention/backends/hpu_attn.py:58]: yes, duplicate this func to use 'is_causal' as False.

File: vllm/worker/hpu_model_runner.py (hunk 4 above)
Comment (xuechendi) [reply vllm/worker/hpu_model_runner.py:1329]: No, I didn't enable that, Yang and I had a discussion on how to add context_length as well, will need to re-work on attn_mask if we do so.

File: vllm/worker/hpu_model_runner.py [hunk 8]
```diff
@@ -1189,6 +1263,241 @@ def _prepare_prompt(
...
+            # where start_idx is max(0, seq_len - sliding_window).
+            # For example, if the prompt len is 10, sliding window is 8, and
+            # block size is 4, the first two tokens are masked and the slot
+            # mapping will be [-1, -1, 2, 3, 4, 5, 6, 7, 0, 1].
+            start_idx = 0
+            if self.sliding_window is not None:
+                assert context_len == 0, (
+                    "Prefix caching is currently not supported with "
+                    "sliding window attention")
+                start_idx = max(0, seq_len - self.sliding_window)
```
[the remark this answers is not in the document]
Comment (xuechendi) [reply vllm/worker/hpu_model_runner.py:1364]: no

File: vllm/worker/hpu_model_runner.py [hunk 9]
```diff
@@ -1542,27 +1853,40 @@ def prepare_input_tensors(
...
-                if paddings_prompt_logprobs else paddings,
+                paddings,
                 dtype=sampling_metadata.selected_token_indices.dtype,
                 device=sampling_metadata.selected_token_indices.device)
-            sampling_metadata.selected_token_indices.add_(paddings)
-        else:
-            sampling_metadata = None
+            sampling_metadata.selected_token_indices = \
+                torch.cat((sampling_metadata.selected_token_indices, paddings),
+                          dim=0)
```
Comment (libinta) [inline vllm/worker/hpu_model_runner.py:1889]: can you add the else portion? for the pooler, there is no sampling_medatdata

File: vllm/worker/hpu_model_runner.py [hunk 10]
```diff
@@ -227,6 +228,33 @@ def find_rope_layer(parent, path):
...
+
+        self.global_state.prompt_buckets = list(
+            filter(
+                lambda bucket: bucket[1] <= origin_max_prompt_len and bucket[0]
+                == 1, prompt_buckets))
+
+        msg = (f"Generated {len(self.global_state.prompt_buckets)} "
+               f"prompt buckets [bs, seq]: "
+               f"{list(sorted(self.global_state.prompt_buckets))}")
+        print(msg)
```
Comment (libinta) [inline vllm/worker/hpu_model_runner.py:255]: this warmup doesn't warm up the seq_lens and context_lens tensors shape in real run. these will be part of the hash in trim_attn_metadata. should warmup include these?

Comment (kamil-kaczor) [reply vllm/worker/hpu_model_runner.py:255]: @libinta I see no "bucket X was not warmup-up" in the logs but in the profling every minute-1.5min there's a decode taking 10x time it should take so perhaps there's some recompilation issue

**Comment by michalkuligowski** [comment]
@kamil-kaczor is this PR still relevant?

**Comment by kamil-kaczor** [comment]
Moving to draft, please for testing use: https://github.com/HabanaAI/vllm-fork/pull/987. Once it's merged I will close this one.

**Comment by kamil-kaczor** [comment]
Merged prefill+ is merged, closing the PR & deleting the branch
